src/bfcli/print.c

```diff
@@ -0,0 +1,176 @@
+{
+    bf_assert(data);
+    // 5 characters per byte (0x%02x) + 1 for the null terminator.
+    char buf[(BF_DUMP_HEXDUMP_LEN * BF_DUMP_TOKEN_LEN) + 1];
+    const void *end = data + len;
+
+    while (data < end) {
+        char *line = buf;
+        for (size_t i = 0; i < BF_DUMP_HEXDUMP_LEN && data < end; ++i, ++data)
+            line += sprintf(line, "0x%02x ", *(unsigned char *)data);
+
+        (void)fprintf(stderr, "%s", buf);
+    }
+}
+
+/**
+ * Dump the details of a chain, including its rules and counters.
+ *
+ * @param chain Pointer to the chain to be dumped. Must be non-NULL.
+ * @param counter Pointer to the array of counters associated with the
+ *        chain. Must be non-NULL if with_counters is true.
+ * @param with_counters Boolean flag indicating whether to include
+ *        counters in the dump.
+ */
+static int bf_cli_chain_dump(struct bf_chain *chain, bf_list *counters,
+                             bool with_counters)
+{
+    struct bf_hook_opts *opts = &chain->hook_opts;
+    struct bf_counter *counter = NULL;
+    uint32_t used_opts = chain->hook_opts.used_opts;
+    bool need_comma = false;
+
+    bf_assert(chain);
+    bf_assert(!with_counters || counters);
+
+    (void)fprintf(stderr, "chain %s", bf_hook_to_str(chain->hook));
+    (void)fprintf(stderr, "{");
+
+    if (used_opts & (1 << BF_HOOK_OPT_ATTACH)) {
+        (void)fprintf(stderr, "attach=%s", opts->attach ? "yes" : "no");
+        need_comma = true;
+    }
+
+    if (used_opts & (1 << BF_HOOK_OPT_IFINDEX)) {
+        if (need_comma)
+            (void)fprintf(stderr, ",");
+        (void)fprintf(stderr, "ifindex=%d", opts->ifindex);
+        need_comma = true;
+    }
+
+    if (used_opts & (1 << BF_HOOK_OPT_NAME)) {
+        if (need_comma)
+            (void)fprintf(stderr, ",");
+        (void)fprintf(stderr, "name=%s", opts->name);
+        need_comma = true;
+    }
+
+    (void)fprintf(stderr, "}");
+    (void)fprintf(stderr, " policy: %s\n", bf_verdict_to_str(chain->policy));
+
+    if (with_counters) {
+        // List order is Error, Policy, and then Rules counters.
+        struct bf_list_node *error_node, *policy_node;
+
+        error_node = bf_list_get_head(counters);
+        if (!error_node)
+            return bf_err_r(-ENOENT, "expected error counter\n");
+
+        policy_node = bf_list_node_next(error_node);
+        if (!policy_node)
+            return bf_err_r(-ENOENT, "expected policy counter\n");
+
+        counter = bf_list_node_get_data(policy_node);
+        (void)fprintf(stderr, "    counters policy %lu packets %lu bytes; ",
+                      counter->packets, counter->bytes);
+        bf_list_delete(counters, policy_node);
+
+        counter = (struct bf_counter *)bf_list_node_get_data(error_node);
+        (void)fprintf(stderr, "error %lu packets %lu bytes \n",
+                      counter->packets, counter->bytes);
+
+        bf_list_delete(counters, bf_list_get_head(counters));
+    }
```

Comment:
You could make it simpler:

```c
struct bf_list_node *counters_node;

counters_node = bf_list_get_head(counters);

// When you want to print counters
counters = bf_list_node_get_data(counters_node);
fprintf(...);
counters_node = bf_list_node_next(counters_node);
```

You don't need to remove the node from the list, the cleanup function will take care of it. Also, you know that your counters are in the order you want to read them, so use `counters_node` as an iterator.

Of course, you need to gate with `if (with_counters)` when necessary.